### PR TITLE
[DPE-2992] - Provide config db to mongos

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
Related mongos charm cannot start without configdb string

## Solution
Provide configdb string to mongos charm

## Testing
```
juju deploy application
juju deploy mongos
juju deploy ./*charm --config role="config-server" config-server-one 
juju deploy ./*charm --config role="shard" shard-one 

# relate
juju integrate mongos application
juju integrate config-server-one:config-server shard-one:sharding
juju integrate mongos:cluster config-server-one:cluster

# relate shards
juju ssh application/0
sudo snap logs charmed-mongodb.mongos
charmed-mongodb.mongosh mongodb://localhost:27018
```

## Future PRs
1. implement this on the mongos charm side
2. basic int test

### CODE REVIEW

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 1]
```diff
@@ -6,19 +6,22 @@
...
 
 logger = logging.getLogger(__name__)
 KEYFILE_KEY = "key-file"
 KEY_FILE = "keyFile"
 HOSTS_KEY = "host"
-CONFIG_SERVER_URI_KEY = "config-server-uri"
+CONFIG_SERVER_DB_KEY = "config-server-db"
 
 # The unique Charmhub library identifier, never change it
 LIBID = "58ad1ccca4974932ba22b97781b9b2a0"
```
Comment (MiaAltieri) [inline lib/charms/mongodb/v0/config_server_interface.py:28]: bump libpatch

File: lib/charms/mongodb/v1/helpers.py [hunk 2]
```diff
@@ -91,14 +92,14 @@ def get_mongos_args(
...
     # mongos running on the config server communicates through localhost
     # use constant for port
-    config_server_uri = f"{config.replset}/localhost:27017"
+    config_server_db = config_server_db or f"{config.replset}/localhost:27017"
 
     full_conf_dir = f"{MONGODB_SNAP_DATA_DIR}{CONF_DIR}" if snap_install else CONF_DIR
     cmd = [
         # mongos on config server side should run on 0.0.0.0 so it can be accessed by other units
         # in the sharded cluster
         "--bind_ip_all",
```
Comment (MiaAltieri) [inline lib/charms/mongodb/v1/helpers.py]: Question

MongoDB has fancy interals so this also sets up a unix socket :
```
ubuntu@juju-8c71f6-1:~$ ss -l | grep -i mongo
u_str LISTEN 0      4096                                      /tmp/mongodb-27018.sock 623618975                       * 0
```

Should we restict it to **only** allow unix socket connections?

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 3]
```diff
@@ -116,35 +132,74 @@ def __init__(
...
             return
 
-        self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        updated_keyfile = self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        updated_config = self.update_config_server_db(
+            config_server_db=relation_data.get(CONFIG_SERVER_DB_KEY)
+        )
+
+        # avoid restarting mongos when possible
+        if not updated_keyfile and not updated_config and self.charm.monogs_initialised:
```
Comment (MiaAltieri) [inline lib/charms/mongodb/v0/config_server_interface.py]: spell check `monogs_initialised `

File: lib/charms/mongodb/v1/helpers.py (hunk 2 above)
Comment (delgod) [reply lib/charms/mongodb/v1/helpers.py]: for config server `--bind_ip_all` is better.
for subordinate charm - only socket

File: lib/charms/mongodb/v0/config_server_interface.py (hunk 3 above)
Comment (delgod) [reply lib/charms/mongodb/v0/config_server_interface.py]: What do we want to check in `mongos_initialised` additionally? I don't remember anything that we can consider as mongos initialisation. it is or started or not started

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 4]
```diff
@@ -116,35 +132,74 @@ def __init__(
...
+
+        # restart on high loaded databases can be very slow (e.g. up to 10-20 minutes).
+        if not self.is_mongos_running():
+            logger.info("mongos has not started, deferring")
+            self.charm.unit.status = WaitingStatus("Waiting for mongos to start")
+            event.defer()
+            return
 
-        # TODO: Follow up PR. Start mongos with the config-server URI
         # TODO: Follow up PR. Add a user for mongos once it has been started
```
Comment (delgod) [inline lib/charms/mongodb/v0/config_server_interface.py:165]: The user should provided by the Provider. Why do we need user creation on the Requirer side?

**Review by delgod — DISMISSED** [review]

File: lib/charms/mongodb/v0/config_server_interface.py (hunk 1 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/config_server_interface.py:28]: hmm, is there should be change in libpatch version?

File: lib/charms/mongodb/v0/config_server_interface.py (hunk 3 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/config_server_interface.py]: Hmm yes, we could check ` if not self.is_mongos_running():`, but this code has a retry in it, so it takes more time to run, perhaps I can disable the retry in certain cases

File: lib/charms/mongodb/v0/config_server_interface.py (hunk 4 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/config_server_interface.py:165]: gotcha

File: lib/charms/mongodb/v0/config_server_interface.py (hunk 1 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v0/config_server_interface.py:28]: yes, whenever code changes the libpatch must be bumped

**Review by delgod — APPROVED** [review]

**Review by dmitry-ratushnyy — APPROVED** [review]

**Pull request merged**